src/VisualStudio/Core/Def/Implementation/ProjectTelemetry/VisualStudioProjectTelemetryService.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.ProjectTelemetry;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.Internal.VisualStudio.Shell;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectTelemetry
+{
+    internal class VisualStudioProjectTelemetryService
+        : ForegroundThreadAffinitizedObject, IProjectTelemetryService, IProjectTelemetryServiceCallback
+    {
+        private const string EventPrefix = "VS/Compilers/Compilation/";
+        private const string PropertyPrefix = "VS.Compilers.Compilation.Inputs.";
+
+        private const string TelemetryEventPath = EventPrefix + "Inputs";
+        private const string TelemetryExceptionEventPath = EventPrefix + "TelemetryUnhandledException";
+
+        private const string TelemetryProjectIdName = PropertyPrefix + "ProjectId";
+        private const string TelemetryProjectGuidName = PropertyPrefix + "ProjectGuid";
+        private const string TelemetryLanguageName = PropertyPrefix + "Language";
+        private const string TelemetryAnalyzerReferencesCountName = PropertyPrefix + "AnalyzerReferences.Count";
+        private const string TelemetryProjectReferencesCountName = PropertyPrefix + "ProjectReferences.Count";
+        private const string TelemetryMetadataReferencesCountName = PropertyPrefix + "MetadataReferences.Count";
+        private const string TelemetryDocumentsCountName = PropertyPrefix + "Documents.Count";
+        private const string TelemetryAdditionalDocumentsCountName = PropertyPrefix + "AdditionalDocuments.Count";
```

Comment:
Should we also track `AnalyzerConfigDocuments.Count`?

Reply:
Sure.  i can add that.  (just preserving existing telemetry code).  That won't break anything, right?

Reply:
Hmm I am not sure - should be fine to do this in a follow-up PR as you are preserving existing semantics.